File: lib/features/audio_visualizer/data/transcription_service.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import '../../../core/supabase/supabase_client.dart' as config;
import 'captions.dart';

/// Chama a edge function `transcribe-audio` (proxy para o serviço Whisper
/// próprio) e devolve as legendas já parseadas.
Future<Captions> transcribeAudio(Uint8List bytes, String mimeType) async {
  final uri = Uri.parse('${config.supabaseUrl}/functions/v1/transcribe-audio');
  final response = await http.post(
    uri,
    headers: {
      'Authorization': 'Bearer ${config.supabaseAnonKey}',
      'apikey': config.supabaseAnonKey,
      'Content-Type': 'application/json',
    },
    body: jsonEncode({
      'audio_base64': base64Encode(bytes),
      'mime_type': mimeType,
    }),
  );

  if (response.statusCode != 200) {
    throw Exception('Erro ${response.statusCode}: ${response.body}');
  }

  return Captions.parse(response.body);
}

File: src/features/audio-visualizer/transcription.service.ts
import { env } from "@/core/env";
import { parseCaptions, type Captions } from "@/features/audio-visualizer/captions";

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  let binary = "";
  const bytes = new Uint8Array(buffer);
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

export async function transcribeAudio(file: File): Promise<Captions> {
  const buffer = await file.arrayBuffer();
  const audioBase64 = arrayBufferToBase64(buffer);

  const response = await fetch(`${env.supabaseUrl}/functions/v1/transcribe-audio`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${env.supabaseAnonKey}`,
      apikey: env.supabaseAnonKey,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ audio_base64: audioBase64, mime_type: file.type }),
  });

  if (!response.ok) {
    throw new Error(`Erro ${response.status}: ${await response.text()}`);
  }

  return parseCaptions(await response.text());
}
